prow/tide/tide.go

```diff
@@ -1022,26 +1029,24 @@ func (c *Controller) trigger(sp subpool, presubmits map[int][]config.Presubmit,
 	// If multiple required jobs have the same context, we assume the
 	// same shard will be run to provide those contexts
 	triggeredContexts := sets.NewString()
-	for _, pr := range prs {
-		for _, ps := range presubmits[int(pr.Number)] {
-			if triggeredContexts.Has(string(ps.Context)) {
-				continue
-			}
-			triggeredContexts.Insert(string(ps.Context))
-			var spec prowapi.ProwJobSpec
-			if len(prs) == 1 {
-				spec = pjutil.PresubmitSpec(ps, refs)
-			} else {
-				spec = pjutil.BatchSpec(ps, refs)
-			}
-			pj := pjutil.NewProwJob(spec, ps.Labels, ps.Annotations)
-			start := time.Now()
-			if _, err := c.prowJobClient.Create(&pj); err != nil {
-				c.logger.WithField("duration", time.Since(start).String()).Debug("Failed to create ProwJob on the cluster.")
-				return fmt.Errorf("failed to create a ProwJob for job: %q, PRs: %v: %v", spec.Job, prNumbers(prs), err)
-			}
-			c.logger.WithField("duration", time.Since(start).String()).Debug("Created ProwJob on the cluster.")
+	for _, ps := range presubmits {
```

Comment:
So technically the previous implementation was not vulnerable to triggering things in `presubmits` that didn't refer to any `pr` -- this impl triggers all presubmits. Do we care? Why bother changing?

Reply:
So we gets changed conceptionally is that now each baseSHA+headSHA(s) combination may have a unique set of presubmits. The result is that we can not view them as decoupled anymore, which is why I essentialy moved the responsibility of picking the right presubmits out of the `trigger` function and then made the `presubmits` arg a slice to properly reflect that.